python/opae.admin/opae/admin/tools/regmap_debugfs.py

```diff
@@ -0,0 +1,163 @@
+import sys
+
+
+class RegmapDebugfs():
+    """RegmapDebugfs is a class for accessing the debugfs
+       inteface of regmaps in the linux kernel"""
+
+    def __init__(self, regmapdir):
+
+        self.registers_path = os.path.join(regmapdir, 'registers')
+        self.range_path = os.path.join(regmapdir, 'range')
+        self.ranges = []
+        self.record_len = 0
+        self.register_stride = 0
+
+    def get_register_info(self):
+        """get information about regmap"""
+        with open(self.registers_path, 'r') as _file:
+            line0 = _file.readline()
+            line1 = _file.readline()
+            self.record_len = len(line0)
+            list0 = line0.strip().split(': ')
+            list1 = line1.strip().split(': ')
+            self.register_stride = int(list1[0], 16) - int(list0[0], 16)
+
+    def get_range_info(self):
+        """get the list of address ranges for the regmap"""
+        with open(self.range_path, 'r') as _file:
+            for line in _file.readlines():
+                list0 = line.strip().split('-')
+                range_dict = {
+                    'min': int(list0[0], 16),
+                    'max': int(list0[1], 16),
+                    'num_records': ((int(list0[1], 16) - int(list0[0], 16)) /
+                                    self.register_stride) + 1,
+                }
+                self.ranges.append(range_dict)
+
+    def reg_read(self, addr):
+        """reg_read returns the value read from the given
+           address in the regmap"""
+
+        found = False
+        num_records = 0
+        for range_dict in self.ranges:
+            if range_dict['min'] <= addr <= range_dict['max']:
+                found = True
+                offset = (addr - range_dict['min'])
+                num_records += offset / self.register_stride
+                break
+            else:
+                num_records += range_dict['num_records']
+
+        if not found:
+            sys.exit("bad read address")
+
+        with open(self.registers_path, 'r') as _file:
+            _file.seek(num_records * self.record_len)
+            line0 = _file.readline()
+
+        list0 = line0.strip().split(': ')
+        raddr = int(list0[0], 16)
+
+        if raddr != addr:
+            msg = "bad addr read back 0x{:x} != 0x{:x}".format(raddr, addr)
+            sys.exit(msg)
+
+        return int(list0[1], 16)
+
+    def reg_write(self, addr, val):
+        """reg_write writes val to the addr in the regmap"""
+
+        for range_dict in self.ranges:
+            if range_dict['min'] <= addr <= range_dict['max']:
+                break
+
+        else:
+            sys.exit("bad write address")
+
+        with open(self.registers_path, 'rb+') as _file:
+            _file.write("{:x} {:x}".format(addr, val).encode())
+
+
+def is_regmap_dir(regmapdir):
+    """check for a valid regmap directory in debugfs"""
+    if not os.path.isdir(regmapdir):
+        sys.exit(f"{regmapdir} is not a valid directory")
```

Comment:
Oh no. You should leave here `raise ArgumentTypeError` because it is the feature of `argparse`. That's the only place where you should have it. Sorry for not being clear.

Reply:
No problem.  I will raise the ArugumentTypeError.